### ISSUE TITLE
Data view: Search & Filtering

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#55100]
Part of https://github.com/WordPress/gutenberg/issues/55083

This ticket is a stub to include designs for search & filtering in tables. Considerations that could affect the design include:

* Scalability – there can be thousands of datapoints that a user might want to filter by. Performance, search, etc. all need to be factored in to the UI.
* Optimise for common use cases: It could be beneficial for some tables to present 'primary' filters to users by default.
* Advanced filter operators. These can be per-filter e.g: "Show posts in categories: recipes **and** reviews" vs. "Show posts in categories: recipes **or** reviews". Or even combined across filters e.g. "Show published posts **and** posts with no comments" vs. "Show published posts **or** posts with no comments". Clearly this can get quite complicated quite quickly, so should receive appropriate prominence.
* Saving filter configurations for reuse.
* Resetting; per-filter and globally.
* Shortcuts; create filter from column headers.
* Loading states while filters are enacted.

## Iterations

Support pages, templates, and template parts pages.

DataViews component

- [ ] Top-level bar:
  - [x] Spacing issues in the top-level bar ([see](https://github.com/WordPress/gutenberg/pull/55992#pullrequestreview-1731806181)). [#56151](https://github.com/WordPress/gutenberg/pull/56151)
  - [ ] Make sure it grows gracefully when there are many of them ([see](https://github.com/WordPress/gutenberg/pull/56468#issuecomment-1823988214)).
  - [x] @oandregal Create a more condensed view for filters in the list layout. [Design](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1847417577). [#56983](https://github.com/WordPress/gutenberg/pull/56983)
  - [ ] Disable filters when there's only one option. E.g. filtering pages by author is pointless on a site with only a single author.
  - [ ] Hide the combobox in search widget when there are more than X options. Probably use a `selectlist`.
- [x] Global search [#54966](https://github.com/WordPress/gutenberg/pull/54966)
- [x] Filter for `ENUMERATION` field. Target [design](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1772753679). @oandregal 
  - [x] Basic implementation based on `SelectControl`. [#55270](https://github.com/WordPress/gutenberg/pull/55270)
  - [x] Users should be able to add/remove filters in the UI. The page should be able to define some default filters. [Design](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1772753679). [#55992](https://github.com/WordPress/gutenberg/pull/55992)
  - [x] Refactor the filter UX to [target design](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1877128587) [#58569](https://github.com/WordPress/gutenberg/pull/58569) @ntsekouras 
  - [x] Create custom component based on `DropdownMenu`. Same UI in top-level & columns. [#56110](https://github.com/WordPress/gutenberg/pull/56110)
  - [x] Allow resetting all filters at once. [Design](https://github.com/WordPress/gutenberg/pull/55508#issuecomment-1794996453). [#55955](https://github.com/WordPress/gutenberg/pull/55955)
  - [x] Implement multiple selection. [#56468](https://github.com/WordPress/gutenberg/pull/56468) [#59610](https://github.com/WordPress/gutenberg/pull/59610) [#59696](https://github.com/WordPress/gutenberg/issues/59696)
  - [x] Implement `NOT IN` operation. [Design](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1798325397). [#56479](https://github.com/WordPress/gutenberg/pull/56479)
  - [x] Fix labels for buttons ([see](https://github.com/WordPress/gutenberg/pull/56110#issuecomment-1814429225)).
  - [x] Support search & scrolling when there are many items.
  - [x] Implement conditions "any" or "all". [Design](https://github.com/WordPress/gutenberg/pull/56479#issuecomment-1824597723).
  - [ ] Logical operators: `AND` & `OR`. Needs design.
- [x] Filter for `DATE` field. Starting [design](https://github.com/WordPress/gutenberg/issues/55100#issuecomment-1814140838).
- [x] Filter for `TEXT` field. Needs design.
- [x] Column filters. [#55508](https://github.com/WordPress/gutenberg/pull/55508)
- [x] Hook for DataViews consumers to easily filter their dataset.
- [ ] https://github.com/WordPress/gutenberg/issues/64898
- [ ] https://github.com/WordPress/gutenberg/issues/64612
- [ ] https://github.com/WordPress/gutenberg/issues/59696

**Comment by jameskoster** [comment]
I shared these designs in https://github.com/WordPress/gutenberg/pull/55508 already, but adding them here for visibility too in case there's any urgent design feedback:

**Filtering**

https://github.com/WordPress/gutenberg/assets/846565/9aa0beff-0e6e-4e80-a206-6baaf37c0624

**Loading & searching filter menus**

https://github.com/WordPress/gutenberg/assets/846565/e272eb2f-3a05-455d-84d3-58e34827c241

**Comment by oandregal** [comment]
@jameskoster I've added a new section for tracking specific tasks.

Thoughts/Questions about some tasks in the backlog:

- For the pages page, what other filters can we implement? It sounds like we could allow filtering by date (the [endpoint](https://developer.wordpress.org/rest-api/reference/pages/#list-pages) allows filtering by date before/after), and also a `NOT IN` filter for authors ("authors not in this list"). Is there a design for any of these?
- Improving the current `IN` filter: it sounds the `DropdownMenu` component to be used is not ready yet https://github.com/WordPress/gutenberg/issues/50459#issuecomment-1741012195 so this could be deferred a bit.
- Show/Hide filters: would users be able to hide the "default" filters for a page? For example, the "pages" page shows author and status filters by default – would the user be able to hide these? If the answer is not, I think we'd need to develop more filters before the show/hide feature makes sense.

I can start working on resetting all filters based on [this design](https://github.com/WordPress/gutenberg/pull/55508#issuecomment-1794996453) while we answer/unblock these questions.

**Comment by jameskoster** [comment]
> For the pages page, what other filters can we implement?

For dates, it would be nice to offer preset options (e.g. 'this week', 'last week', 'this month', 'last month'), plus the ability to select a custom date range via datepicker. Would that be feasible? If so I'll work on a design.

Latest designs for the `NOT IN` operator make use of a 'Settings' flyout on the filter menu:

<img width="1238" alt="Filter settings" src="https://github.com/WordPress/gutenberg/assets/846565/84f2fe33-6be5-4a1d-a4db-d57e142eee4c">

As you said, we may have to defer until the component is ready. 

In terms of what else to filter by; `discussion settings` and `template` spring to mind, but they're lower priority. Template filtering would be multi select and require an 'any'/'all' setting.

> would users be able to hide the "default" filters for a page?

Default filters should not be hide-able, but should be reset-able when a non-default value/setting is selected. This can be done locally via the "Reset" action in the menu (see screenshot above), or globally via the reset button in the main filter bar (the design you linked to).

**Comment by oandregal** [comment]
> For dates, it would be nice to offer preset options (e.g. 'this week', 'last week', 'this month', 'last month'), plus the ability to select a custom date range via datepicker. Would that be feasible? If so I'll work on a design.

Yeah, absolutely. We need to figure out the preset => before/after conversion, but it seems doable.

Looking at the existing components, I see a [DatePicker](https://wordpress.github.io/gutenberg/?path=/docs/components-datepicker--docs), a [TimePicker](https://wordpress.github.io/gutenberg/?path=/docs/components-timepicker--docs) and a [DateTimePicker](https://wordpress.github.io/gutenberg/?path=/docs/components-datetimepicker--docs), but they do not work with ranges – the sooner we have a design, the better, so we can update them / create new ones as necessary.

I am also thinking that date presets need to be contextual to the dataset. There may be scenarios where shorter ranges are useful ('last hour' or 'today'), like for a store or a news site. Initially, we could offer some that cover a wide surface, and look at making them configurable per page/dataset if necessary.

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/55270 implements the ability to show/hide filters.

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/56110 starts implementing the new filter component.

**Comment by jameskoster** [comment]
Just a note to say that the datepicker design might take some time to get right, given all the permutations. 

Until then, perhaps for date filtering we simply seek parity with wp-admin, IE filter by month. Hopefully that's simple to do with the current components?

**Comment by oandregal** [comment]
PR for implementing multi-selection in filter https://github.com/WordPress/gutenberg/pull/56468

(also updated a bit the issue description with latest progress)

**Comment by oandregal** [comment]
PR implementing `NOT IN` operator https://github.com/WordPress/gutenberg/pull/56479

**Comment by jameskoster** [comment]
Following on from the recent [updates to the List layout](https://github.com/WordPress/gutenberg/pull/56746), we need to reduce the width of the data view column. But before we can do that, we need a solution for displaying and managing filters in smaller areas. Filter labels can get quite long, and while wrapping is okay, it might not be the best solution:

<img width="1640" alt="filters-wrapping" src="https://github.com/WordPress/gutenberg/assets/846565/6f1da200-d5ac-485f-89b0-4cc8f44a8702">

An alternative approach could be to hide the filter dropdowns in smaller areas, and make filters manageable in the main filter menu:

<img width="2710" alt="filters added" src="https://github.com/WordPress/gutenberg/assets/846565/4bd5a16e-597c-4243-ae2c-5bfbb5d3560f">

Note how the Author filter remains visible in the menu even after it's been added. 

In List layout, and other layouts on mobile devices this would work like so:

<img width="1640" alt="filters-small-spaces" src="https://github.com/WordPress/gutenberg/assets/846565/f5013a5c-4438-4281-ac0f-8823afca377b">

In the absence of the filter dropdowns, a small count ensures that it's easy to tell at a glance whether filters have been applied.

**Comment by oandregal** [comment]
First step to offer a more condense filter affordance at https://github.com/WordPress/gutenberg/pull/56983

**Comment by jameskoster** [comment]
Here's an alternative exploration for the general filter UX. Instead of shoe-horning everything into a DropdownMenu component with multiple flyouts, a single menu instance is used to add filters, and a popover instance contains the filter configuration.

This approach affords us more flexibility in terms of design which will be important when we add more operator options, and other types of filter such as text or date.

As demonstrated in the screenshot below, as soon as you elect to add a filter the popover is opened and focus transfers to the relevant element inside:

<img width="1890" alt="Screenshot 2024-01-04 at 13 25 40" src="https://github.com/WordPress/gutenberg/assets/846565/b263c2c3-6edf-4505-983c-e4469d8b445d">

The operator is controlled via `SelectControl` at the top of the popover:

<img width="1018" alt="Screenshot 2024-01-04 at 13 28 17" src="https://github.com/WordPress/gutenberg/assets/846565/cd54e8b8-1342-4b96-a48d-0a0704028218">

As a rough mockup, here's how this pattern could be applied to a title and date filter UI:

<img width="1190" alt="Screenshot 2024-01-04 at 13 48 38" src="https://github.com/WordPress/gutenberg/assets/846565/ad205a99-ef06-4b07-993e-695efdbaf717">

Narrow containers are still a bit of a conundrum, but there are options to explore there such as horizontal scrolling, or truncation:
<img width="1502" alt="Screenshot 2024-01-04 at 13 51 09" src="https://github.com/WordPress/gutenberg/assets/846565/1055224f-c777-4e81-9eed-61fbd93df4f3">

**Comment by ciampo** [comment]
IMO this is a better UX. It also allows us to have better semantics, by keeping menu stuff in `DropdownMenu`, and having the filter UI in a separate popover.

We'll still probably have to go with a "custom" implementation for the popover, which will include:

- a `listbox` (ie. `SelectControl`) for picking the condition
- a custom data-picking widget (lika a `combobox` for text values, a date picker for date values, ...)
- potentially extra buttons (ie. to reset the filter)

Curious to hear @andrewhayward and @diegohaz 's thoughts too.

**Comment by andrewhayward** [comment]
> Here's an alternative exploration for the general filter UX. Instead of shoe-horning everything into a DropdownMenu component with multiple flyouts, a single menu instance is used to add filters, and a popover instance contains the filter configuration.

This is a much improved direction, from a UX perspective; I was starting to get a bit worried that everything was going to be a menu!

We'll need to think a bit about how users interact with these, at a high level. For example, what happens if you <kbd>+ Add filter</kbd> then immediately hit `Escape`? Does the filter collapse in place, having no effect? Or do we remove it, open the original menu again, and focus the appropriate menu item?

Similarly, what happens if the filter is already active and the user hits `Escape`? Does it undo any changes since you opened the filter popup? Or does it just collapse the popup, with the expectation that those changes are "live"?

The same applies for <kbd>Reset</kbd> too, but that's an even more explicit action. What does it reset the filter to?

**Comment by andrewhayward** [comment]
> We'll still probably have to go with a "custom" implementation for the popover, which will include:
> 
> * a `listbox` (ie. `SelectControl`) for picking the condition
> * a custom data-picking widget (lika a `combobox` for text values, a date picker for date values, ...)
> * potentially extra buttons (ie. to reset the filter)

What @ciampo said seems about right to me; essentially a non-modal dialog positioned next to the relevant filter button.

We should be able to make a base filter implementation that handles the basics, and ensures a consistent experience; something like this brain-dump of pseudo-ish code...

```jsx
function Filter ( props ) {
  const focusRef = useRef();
  const [ isExpanded, setIsExpanded ] = useState( false );
  const toggleIsExpanded = () => setIsExpanded(( expanded ) => ! expanded );
  const { title, activeCondition, values } = props;

  useEffect(() => {
    if ( isExpanded && focusRef.current ) {
      focusRef.current.focus();
    }
  }, [ isExpanded, focusRef ] );

  return (
    <PopoverButton isExpanded={ isExpanded } onClick={ toggleIsExpanded }>
      { `${ title } ${ activeCondition }: ${ values }` }
      { isExpanded && (
        <FilterPopover { ...props } ref={ focusRef } />
      ) }
    </PopoverButton>
  );
}

const FilterPopover = forwardRef(
  function FilterPopover ( { render, ...props }, ref ) {
    const { title, conditions, activeCondition } = props;
    return (
      <Popover>
        { title } <Select values={ conditions } selectedValue={ activeCondition } />
        { render( props, ref ) }
        <Button onClick={ ... }>Reset</Button>
      </Popover>
    );
  }
);
```

```jsx
function TextFilter ( props ) {
  return (
    <Filter { ...props } render={ ( props, ref ) => (
      <TextInput { ...props } ref={ ref } />
    ) } />
  );
}

function ListFilter ( props ) {
  return (
    <Filter { ...props } render={ ( props, ref ) => (
      <ComboBox { ...props } ref={ ref } />
    ) } />
  );
}

function DateFilter ( props ) {
  return (
    <Filter { ...props } render={ ( props, ref ) => (
      <DatePicker { ...props } ref={ ref } />
    ) } />
  );
}
```

We'll also have think about how translation works here; for example, `${ title } ${ condition }` (_<Title\> <starts with\>_, _<Category\> <is\>_, etc) is fine in English, but might not be in other languages.

**Comment by jameskoster** [comment]
> For example, what happens if you + Add filter then immediately hit Escape? Does the filter collapse in place, having no effect? Or do we remove it, open the original menu again, and focus the appropriate menu item?

Good question. On the surface the latter seems more helpful, but if you're someone who wants to add filters before configuring them it might be unexpected/frustrating? 

Perhaps it's best (and probably simplest) to start with consistency. So regardless of whether you just added the filter, or added it previously an re-opened the configuration popover, pressing escape would close the popover and transfer focus to the trigger. 

As for saving, the changes would occur "live", so in effect there's nothing to discard or apply on close.

**Comment by SaxonF** [comment]
+1 on this approach @jameskoster . It also aligns better with the idea of primary filters that are always visible for quick access but may be empty. We could still condense into a single dropdown on smaller containers.

**Comment by ntsekouras** [comment]
I've started working on the redesign and hopefully very soon will have a PR.

**Comment by jameskoster** [comment]
For the filter buttons / chips we might try something like:

<img width="634" alt="Chips" src="https://github.com/WordPress/gutenberg/assets/846565/aa818779-9509-4fa4-964f-a7c51b2684c1">

This would produce the following:

<img width="1424" alt="Screenshot 2024-02-06 at 14 44 09" src="https://github.com/WordPress/gutenberg/assets/846565/4460fd6b-2d94-4a7e-8115-8f6a62e6afe8">

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/59610 implements multi-selection for filters.

**Comment by oandregal** [comment]
@jameskoster is there any design for filters of type "open text"? This is, filters that don't have a preselection of values. At the moment, this is done via the global search, though we may want to have a dedicated filter for that with support for operators such as "Starts with...", "Ends with...", "Contains..." (the current global search).

**Comment by jameskoster** [comment]
Yes!

<img width="789" alt="Screenshot 2024-03-12 at 14 50 39" src="https://github.com/WordPress/gutenberg/assets/846565/f1dd04f0-746c-41b8-8f20-fabaa6258105">

It would be good to get thoughts on the operators here. We could potentially start simple ("Contains" / "Does not contain") and expand from there. What do you think?

**Comment by oandregal** [comment]
What's the relation between this new filter and the "global search"? The global search is quite opaque in that the user doesn't know the fields or operators that it uses. I wonder if both should be connected.

**Comment by jameskoster** [comment]
Good question. The global search currently looks for matches in title and content, is that right? That seems like a fairly natural default behavior to me, for now at least. Perhaps in the future it can support filtering syntax, and reflect filters added manually, a bit like the search on github issues.

<img width="723" alt="Screenshot 2024-03-12 at 15 44 09" src="https://github.com/WordPress/gutenberg/assets/846565/19f5127e-f5b5-4972-b0a8-4af800f071f4">

That would need some more design exploration though, and I'd welcome ideas!

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/pull/59953 adds support for `AND` operators — there's no current use case in pages, templates, parts, and patterns though it's needed in future post types (post, media).

**Comment by youknowriad** [comment]
I'm going to close this issue for now. I think we can explore more smaller/dedicated issues when it comes to it.
